src/ngraph/codegen/file_util.cpp

```diff
@@ -0,0 +1,349 @@
+        if (path != nullptr)
+        {
+            break;
+        }
+    }
+    if (path == nullptr)
+    {
+        path = "/tmp";
+    }
+
+    return path;
+}
+
+vector<char> codegen::file_util::read_file_contents(const string& path)
+{
+    size_t file_size = get_file_size(path);
+    vector<char> data;
+    data.reserve(file_size);
+    data.resize(file_size);
+
+    FILE* f = fopen(path.c_str(), "rb");
+    if (f)
+    {
+        char* p = data.data();
+        size_t remainder = file_size;
+        size_t offset = 0;
+        while (f && remainder > 0)
+        {
+            size_t rc = fread(&p[offset], 1, remainder, f);
+            offset += rc;
+            remainder -= rc;
+        }
+        fclose(f);
+    }
+    else
+    {
+        throw std::runtime_error("error opening file '" + path + "'");
+    }
+    return data;
+}
+
+std::string codegen::file_util::read_file_to_string(const std::string& path)
+{
+    std::ifstream f(path);
+    std::stringstream ss;
+    ss << f.rdbuf();
+    return ss.str();
+}
+
+void codegen::file_util::iterate_files(const string& path,
+                                      std::function<void(const string& file, bool is_dir)> func,
+                                      bool recurse)
+{
+    vector<string> files;
+    vector<string> dirs;
+    file_util::iterate_files_worker(path,
+                                    [&files, &dirs](const string& file, bool is_dir) {
+                                        if (is_dir)
+                                            dirs.push_back(file);
+                                        else
+                                            files.push_back(file);
+                                    },
+                                    recurse);
+
+    for (auto f : files)
+    {
+        func(f, false);
+    }
+    for (auto f : dirs)
+    {
+        func(f, true);
+    }
+}
+
+void codegen::file_util::iterate_files_worker(
+    const string& path, std::function<void(const string& file, bool is_dir)> func, bool recurse)
+{
+    DIR* dir;
+    struct dirent* ent;
+    if ((dir = opendir(path.c_str())) != nullptr)
+    {
+        while ((ent = readdir(dir)) != nullptr)
+        {
+            string name = ent->d_name;
+            switch (ent->d_type)
+            {
+            case DT_DIR:
+                if (name != "." && name != "..")
+                {
+                    string dir_path = file_util::path_join(path, name);
+                    if (recurse)
+                    {
+                        iterate_files(dir_path, func, recurse);
+                    }
+                    func(dir_path, true);
+                }
+                break;
+            case DT_LNK: break;
+            case DT_REG:
+            {
+                string file_name = file_util::path_join(path, name);
+                func(file_name, false);
+                break;
+            }
+            default: break;
+            }
+        }
+        closedir(dir);
```

Comment:
what if `func` throws? you'll leak the pointer?  should `func` be `noexcept`?

Reply:
My understanding of noexcept is that it is meaningless. A noexcept function can throw exceptions.

Reply:
ok, so it needs a guard then?

Reply:
yes it does.